wdl/downsampling_pipeline_part2.wdl

```diff
@@ -1,534 +1,638 @@
-            I=~{bam_sorted_rg_file} \
-            O=~{bam_markdup_name} \
-            M=~{markdup_metrics_name}
-
-        #bam to cram
-        samtools view \
-                -C \
-                -T ~{reference_fasta} \
-                ~{bam_markdup_name} \
-                > ~{cram_downsample_name}
-    >>>
-
-    runtime {
-        docker: downsample_docker
-        memory: mem_size
-        cpu: "1"
-        disks: "local-disk " + disk_size + " HDD"
-        preemptible: 3
-        maxRetries: 1
-    }
+  output {
+    File cram_downsample_file = cram_downsample_name
+    File markdup_metrics_file = markdup_metrics_name
+  }
+
+  command <<<
+    set -euo pipefail
+    #Identifies duplicate reads.
+    picard MarkDuplicates \
+      I=~{bam_sorted_rg_file} \
+      O=~{bam_markdup_name} \
+      M=~{markdup_metrics_name}
+
+    #bam to cram
+    samtools view \
+      -C \
+      -T ~{reference_fasta} \
+      ~{bam_markdup_name} \
+      > ~{cram_downsample_name}
+  >>>
+
+  runtime {
+    docker: downsample_docker
+    cpu: select_first([runtime_attr.cpu_cores, default_attr.cpu_cores])
+    memory: select_first([runtime_attr.mem_gb, default_attr.mem_gb]) + " GiB"
+    disks: "local-disk " + select_first([runtime_attr.disk_gb, default_attr.disk_gb]) + " HDD"
+    bootDiskSizeGb: select_first([runtime_attr.boot_disk_gb, default_attr.boot_disk_gb])
+    preemptible: select_first([runtime_attr.preemptible_tries, default_attr.preemptible_tries])
+    maxRetries: select_first([runtime_attr.max_retries, default_attr.max_retries])
+  }
 }
 
 task sortIndex {
     
-    input {
-
-        File cram_downsample_file
-        String downsample_docker
-        Int disk_size
-        String mem_size
-    }
-
-    String downsample_file_sorted_name = basename(cram_downsample_file, ".cram") + "_sorted.cram"
+  input {
+    File cram_downsample_file
+    String downsample_docker
+    RuntimeAttr? runtime_attr_override
+  }
+
+  Int num_cpu = 1
+  Int mem_size_gb = 4
+  Int vm_disk_size = 50
+
+  RuntimeAttr default_attr = object {
+    cpu_cores: num_cpu,
+    mem_gb: mem_size_gb, 
+    disk_gb: vm_disk_size,
+    boot_disk_gb: 20,
+    preemptible_tries: 3,
+    max_retries: 1
+  }
+
+  RuntimeAttr runtime_attr = select_first([runtime_attr_override, default_attr])
+
+  String downsample_file_sorted_name = basename(cram_downsample_file, ".cram") + "_sorted.cram"
     
-    output {
-        File cram_sorted_file = downsample_file_sorted_name
-        File crai_file = downsample_file_sorted_name + ".crai"
-    }
+  output {
+    File cram_sorted_file = downsample_file_sorted_name
+    File crai_file = downsample_file_sorted_name + ".crai"
+  }
 
-    command <<<
+  command <<<
+      
+    set -euo pipefail
+      
+    #sorts and indexes the downsampled, markduped cram file
 
-        samtools sort -o ~{downsample_file_sorted_name} ~{cram_downsample_file}
+    samtools sort -o ~{downsample_file_sorted_name} ~{cram_downsample_file}
```

Comment:
I notice that samtools sort was called in a previous task - just checking that this needs to be done twice? It looks like it's needed before markDuplicates, and then does markDuplicates / bamToCram cause it to no longer be sorted?

Reply:
This might be redundant but I am not 100% certain, I googled and it seems to maintain order and I just need to index, but not 100% certain

Reply:
You have these intermediate files on ErisOne right? Maybe you could just check if they're identical before and after sorting